File: 1 data_structures/hash_table/hash_table.cs
using System;
using System.Collections.Generic;

class HashTable
{
    private Dictionary<int, List<KeyValuePair<string, int>>> buckets;

    public HashTable()
    {
        buckets = new Dictionary<int, List<KeyValuePair<string, int>>>();
    }

    private int GetBucketIndex(string key)
    {
        return Math.Abs(key.GetHashCode() % buckets.Count);
    }

    public void Add(string key, int value)
    {
        int index = GetBucketIndex(key);
        if (!buckets.ContainsKey(index))
        {
            buckets[index] = new List<KeyValuePair<string, int>>();
        }
        else
        {
            // Check if key already exists in the bucket
            foreach (var pair in buckets[index])
            {
                if (pair.Key == key)
                {
                    throw new ArgumentException("Key already exists");
                }
            }
        }

        // Add key-value pair to the bucket
        buckets[index].Add(new KeyValuePair<string, int>(key, value));
    }

    public int Get(string key)
    {
        int index = GetBucketIndex(key);
        if (buckets.ContainsKey(index))
        {
            foreach (var pair in buckets[index])
            {
                if (pair.Key == key)
                {
                    return pair.Value;
                }
            }
        }
        throw new KeyNotFoundException("Key not found");
    }

    public void Remove(string key)
    {
        int index = GetBucketIndex(key);
        if (buckets.ContainsKey(index))
        {
            for (int i = 0; i < buckets[index].Count; i++)
            {
                if (buckets[index][i].Key == key)
                {
                    buckets[index].RemoveAt(i);
                    return;
                }
            }
        }
        throw new KeyNotFoundException("Key not found");
    }
}

class Program
{
    static void Main(string[] args)
    {
        HashTable table = new HashTable();
        table.Add("apple", 5);
        table.Add("banana", 7);
        table.Add("orange", 3);

        Console.WriteLine("Value for 'apple': " + table.Get("apple"));
        Console.WriteLine("Value for 'banana': " + table.Get("banana"));
        Console.WriteLine("Value for 'orange': " + table.Get("orange"));

        table.Remove("apple");

        try
        {
            Console.WriteLine("Value for 'apple': " + table.Get("apple"));
        }
        catch (KeyNotFoundException)
        {
            Console.WriteLine("Key 'apple' not found");
        }
    }
}

File: algorithms/1 data_structures/hash_table/hash_table.py
class HashTable:
    def __init__(self):
        self.buckets = {}

    def _get_bucket_index(self, key):
        return hash(key) % len(self.buckets)

    def add(self, key, value):
        index = self._get_bucket_index(key)
        if index not in self.buckets:
            self.buckets[index] = []
        else:
            # Check if key already exists in the bucket
            for k, _ in self.buckets[index]:
                if k == key:
                    raise ValueError("Key already exists")

        # Add key-value pair to the bucket
        self.buckets[index].append((key, value))

    def get(self, key):
        index = self._get_bucket_index(key)
        if index in self.buckets:
            for k, v in self.buckets[index]:
                if k == key:
                    return v
        raise KeyError("Key not found")

    def remove(self, key):
        index = self._get_bucket_index(key)
        if index in self.buckets:
            for i, (k, v) in enumerate(self.buckets[index]):
                if k == key:
                    del self.buckets[index][i]
                    return
        raise KeyError("Key not found")

table = HashTable()
table.add("apple", 5)
table.add("banana", 7)
table.add("orange", 3)

print("Value for 'apple':", table.get("apple"))
print("Value for 'banana':", table.get("banana"))
print("Value for 'orange':", table.get("orange"))

table.remove("apple")
print("Value for 'apple':", table.get("apple"))  # This will raise KeyError